pkg/reconciler/knativeserving/ingress/ingress.go

```diff
@@ -87,61 +49,90 @@ func Transformers(ctx context.Context, ks *v1beta1.KnativeServing) []mf.Transfor
 	return transformers
 }
 
-func getIngress(version string) (mf.Manifest, error) {
-	// If we can not determine the version, append no ingress manifest.
-	if version == "" {
+func getIngress(path string) (mf.Manifest, error) {
+	if path == "" {
 		return mf.Manifest{}, nil
 	}
+	return common.FetchManifest(path)
+}
+
+func getIngressPath(version string, ks *v1beta1.KnativeServing) string {
+	var urls []string
 	koDataDir := os.Getenv(common.KoEnvKey)
-	// Ingresses are saved in the directory named major.minor. We remove the patch number.
-	ingressVersion := common.LATEST_VERSION
+	sourceVersion := common.LATEST_VERSION
 	if !strings.EqualFold(version, common.LATEST_VERSION) {
-		ingressVersion = semver.MajorMinor(common.SanitizeSemver(version))[1:]
+		sourceVersion = semver.MajorMinor(common.SanitizeSemver(version))[1:]
+	}
+
+	// This line can make sure a valid available source version is returned.
+	ingressPath := filepath.Join(koDataDir, "ingress", sourceVersion)
+	if ks.Spec.Ingress == nil {
+		url := filepath.Join(ingressPath, "istio")
+		urls = append(urls, url)
+		return strings.Join(urls, common.COMMA)
+	}
+
+	if ks.Spec.Ingress.Istio.Enabled {
+		url := filepath.Join(ingressPath, "istio")
+		urls = append(urls, url)
+	}
+	if ks.Spec.Ingress.Contour.Enabled {
+		url := filepath.Join(ingressPath, "contour")
+		urls = append(urls, url)
+	}
+	if ks.Spec.Ingress.Kourier.Enabled {
+		url := filepath.Join(ingressPath, "kourier")
+		urls = append(urls, url)
 	}
 
-	// This line can make sure a valid available ingress version is returned.
-	ingressVersion = common.GetLatestIngressRelease(ingressVersion)
-	ingressPath := filepath.Join(koDataDir, "ingress", ingressVersion)
-	return common.FetchManifest(ingressPath)
+	if len(urls) == 0 {
+		url := filepath.Join(ingressPath, "istio")
```

Comment:
We have the default behavior around L#67 so it should be fine, I think. But currently (without this PR) `nonFilter` was returned and no ingress was created. So, it should be same behavior instead of installing Istio?

Reply:
Right. Let me change it. I can add some more test cases to see if the edge cases are covered.